### DISCUSSION TITLE
A way to exclude invalid items?

### DISCUSSION THREAD

**Discussion by kladnik** [colinhacks/zod#672] [Q&A] [answered]
Hi!

I am having trouble with solving an issue.
Is there a way to exclude the items that are not valid while not returning the error?
If we have an array of strings and one of them is `null` I'd like it to return the cleaned version of the array.

```ts
const array = ['test', 'test2', 'test3', null]
// Wanted behavior
const validatedArray = z.array(z.string()).parse(array) // ['test', 'test2', 'test3']
```

With very complex types this would be very practical. Currently I am looping over each item in array and checking, which seems very complicated especially once you're nested deeply.

Is there any better solution?

**Comment by scotttrinh** [comment, marked as answer]
I'm not aware of a way to adjust a Zod parser's behavior to get it to simply ignore items rather than throwing an error. You could write an array helper function and use `preprocess` maybe, but it's not going to be pretty since `preprocess` starts at `unknown` and requires you to do your own type narrowing.

If I was trying to implement a solution here, I would specify all of the types, or use `z.array(z.any())` and use transform like this:

```ts
const array = ['test', 'test2', 'test3', null]
const validatedArray = z.array(z.any())
  .transform(
    (as) => as.filter((a) => z.string().safeParse(a).success)
  )
  .parse(array) // ['test', 'test2', 'test3']
```

Maybe this is a useful generalization?
[CodeSandbox](https://codesandbox.io/s/youthful-butterfly-36b0t?file=/src/index.ts)

```ts
const makeFilteredSchema = (s: z.ZodTypeAny) =>
  z.array(z.any()).transform(
    (as) => as.filter((a) => s.safeParse(a).success)
  );

const filteredStringArraySchema = makeFilteredSchema(z.string());
const filteredStringOrNumberSchema = makeFilteredSchema(z.union([z.string(), z.number()]);

// More likely used inline:

const fooSchema = z.object({
  bar: makeFilteredSchema(z.string()),
});
```

**Reply by kladnik** [reply]
Thanks! I did manage to do something similar with `preprocess`, not super pretty, but it works.
I do think a new option `.exclude()` on arrays and optional fields would be super practical though.

**Reply by scotttrinh** [reply]
Having truly heterogeneous arrays always feels like a smell to me, so I'm tempted to push back on the idea of including it as a part of the core library. If your array can contain `string | null`, I'd rather see the input schema reflect that and then use `transform` or a downstream filter to remove `null` or unexpected values. I think this keeps us a bit more aligned with TypeScript's concept of the `Array` type. But, you can certainly wrap Zod and include a `ZodFilteredArray` type as a custom type if you find a lot of use for it!

**Reply by samarshrestha720** [reply]
Hi, the accepted answer didn't work on my deeply nested array `{field1: string, valuesTypeA: [...values], valuesTypeB:[...values]}` . Could you share your way with the preprocess?

**Reply by iiAku** [reply]
@samarshrestha720 I've done a preprocess version here, don't know if it can help you

https://github.com/colinhacks/zod/discussions/672#discussioncomment-11878465

**Comment by flofehrenbacher** [comment]
Hello everybody, check out my `preprocess` solution with type narrowing -> https://gist.github.com/flofehrenbacher/71f93f7a423cfc11e1dd018325c241ec
Feedback is highly appreciated 🤓

**Comment by aurelticot** [comment]
I made a couple of tweaks to @scotttrinh's answer

```ts
const makeFilteredArraySchema = <T extends ZodSchema>(schema: T) =>
  z
    .array(z.unknown()) // `z.unknown()` more relevant than `z.any()` and avoid linting issue with any
    .transform((items) =>
      items?.filter(
        (item): item is z.infer<T> => schema.safeParse(item).success
        // `item is z.infer<T>` prevents the output type to be `unknown`. Ok to cast the type as it's been validated
      )
    );

const fooSchema = z.object({
  bar: makeFilteredArraySchema(z.string()),
});

const fooType = z.infer<typeof fooSchema>
// fooType = { bar: string[] } 
```

**Reply by SharmaTushar** [reply]
@aurelticot The return type doesn't seem be correctly inferred in one of my cases

```ts
import { z } from 'zod'
import { makeFilteredArraySchema } from './utils'

const menuBaseVariation = z.object({
    id: z.string(),
    title: z.string(),
})

const baseMenuItemSchema = z.object({
    model: z.string(),
    variations: menuBaseVariation.array(),
})

export type MenuItem = z.infer<typeof baseMenuItemSchema>['variations'][number] &
    (
        | { links?: MenuItem[] }
        | {
              type: string
              link: string
              opentype: string
          }
    )

export type MenuItemRaw = {
    model: string
    variations: ({
        id: string
        title: string
    } & (
        | { links?: MenuItemRaw[] }
        | {
              type: string
              link: string
              opentype?: string
          }
    ))[]
}

export const menuItemSchema: z.ZodType<MenuItem, z.ZodTypeDef, MenuItemRaw> = baseMenuItemSchema
    .extend({
        variations: makeFilteredArraySchema(
            z.union([
                menuBaseVariation.extend({
                    links: z.lazy(() => makeFilteredArraySchema(menuItemSchema).optional()),
                }),
                menuBaseVariation.extend({
                    type: z.string(),
                    link: z.string(),
                    opentype: z.string().optional(),
                }),
            ]),
        ),
    })
    .transform((item) => item.variations[0]!)
```

so I had to force TS to ignore it
```ts
export const makeFilteredArraySchema = <T extends z.ZodSchema>(schema: T): ReturnType<typeof z.array<T>> => {
    // @ts-expect-error
    return z
        .array(z.unknown()) // `z.unknown()` more relevant than `z.any()` and avoid linting issue with any
        .transform((items) => items?.filter((item): item is z.infer<T> => schema.safeParse(item).success))
}
```

Now for some reason, it's not correctly transforming the lazy items and instead of getting `item.variations[0]`, I'm simply getting the `item` itself. Maybe the order of transformations has changed. I wonder how to preserve it.

*Edit:*
Using just filter transforms, but doesn't use the transformed result. A map step before filter has fixed it, maybe can be optimized to avoid looping over twice.

```ts
export const makeFilteredArraySchema = <T extends z.ZodSchema>(schema: T) => {
    return z
        .array(z.unknown()) // `z.unknown()` more relevant than `z.any()` and avoid linting issue with any
        .transform(
            (items) =>
                items
                    ?.map((item): z.infer<T> | undefined => {
                        const result = schema.safeParse(item)

                        if (result.success) return result.data
                        return undefined
                    })
                    .filter((item): item is z.infer<T> => item !== undefined),
        )
}
``` 

*Edit:*

I spoke too soon.
![image](https://github.com/colinhacks/zod/assets/6995887/7e2fca68-06fc-4a75-86fb-d5b05602fe1b)

safeParse result has success `true` and data is `undefined`. Is this supposed to be happening?

**Comment by varunpvp** [comment]
This might help.
```typescript
const zodArray = <S>(itemSchema: z.ZodType<S>) => {
  const catchValue = {} as never;

  return z
    .array(itemSchema.catch(catchValue))
    .transform((a) => a.filter((o) => o !== catchValue))
    .catch([]);
};
```
Example
```typescript
const personSchema = z.object({
  name: z.string(),
  phones: zodArray(z.object({ label: z.string(), number: z.string() }))
});

const person = {
  name: "John",
  phones: [
    { number: "1234567890" }, // skipped
    { label: "Home" }, // skipped
    {}, // skipped
    { label: "Mobile", number: "1234567891" }
  ]
};
```
Output
```json
{
  "success": true,
  "data": {
    "name": "John",
    "phones": [
      {
        "label": "Mobile",
        "number": "1234567891"
      }
    ]
  }
}
```

**Comment by lifeiscontent** [comment]
this seems to be the most complete implementation I could muster up:

```ts
const makeFilteredSchema = <S extends z.ZodTypeAny>(s: S) => {
  return z.array(z.unknown()).transform((as) => {
    const result: S[] = [];
    for (const a of as) {
      const parsed = s.safeParse(a);
      if (parsed.success) {
        result.push(parsed.data);
      }
    }
    return result;
  });
};
```

the reason why filter is not a great choice here, is because the schema itself could have additional transforms or maybe the data would have additional keys and in the case of filter they would not be removed, but because we're individually parsing each item, those additional transforms would be preserved.

**Reply by sorenhoyer** [reply]
@lifeiscontent I think it would still be relevant to many to be able to log the erroneous items that got removed, as zod issues ideally.

That seems to be the only thing missing.

**Reply by lifeiscontent** [reply]
@sorenhoyer can you give details in how you would use what you're expressing? I don't understand the use case.

**Reply by iiAku** [reply]
I believe what @sorenhoyer meant was a way to log out (at a certain level) what element are being filtered out from initial (pre-validation) list of items.

This solution works to me, I'm just struggling about the infered type when `makeFilteredSchema` is existing as part of a schema eg:

```
const mySchema = z.object({
  a: z.string(),
  b: makeFilteredSchema(anotherSchema)
})
```
Infered values of `mySchema` will return unknown due to `z.array(unknown)`. Don't know how to make it infered properly.

**Reply by lifeiscontent** [reply]
@iiAku can you provide a minimal complete example? I can try to help but `anotherSchema` is hard to infer 😅 

here's a basic idea of how to use makeFilteredSchema:

https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcBEyEAJvgNwBQFAxhAHYDO8IAhgNYCmAYsADYwcoHYgGVqACw6s4AXjgAeEXA4APAXWIMUAOgBaJACoBPMBwCCdIwD4AFAwBccEQEpZVxBThwhMAK5Q6HRYoKBYjG2RtXzo2OggAdzobZ2dtGFDGTGgQGxsWBlcZdwRPLzhaRnghBl9+RxEAbQBdWThmyjKsaDgbCqY4FjgITAGCj06vPvgwYIZhVoZtBhZMDgAFWY485w6J4BGbGag54iXfamoOBjGSibLq2phtMF8GcUPN0+IWGBYd0s6aAB6GBPn8gQe-F2aH+aEoNHo-QYEikgzkkQgACMAFYcagwGy3ZQhaCOSJMKDAOgAc2SABpSpCYA44KxODx+IJhGJJKwItpgqFwhicXiCUhVCxwLwOGSlukqbTXDCUgyYfCpl0IAsUaxnptCeqKEA

**Reply by iiAku** [reply]
I come up with a slightly different version (preprocess based) of yours and (at least to me) it's seems to work and infer better.

The problem I had was the inference were fine from the type when you create it from the schema even with your version.

However if you mimic what you would had with raw TS type and match it against that TS type is does not seems to be equal.

Here an example showing off what I am saying [link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFA9NXAMZQCmxwMAzljngDbCZNg7ehAB2MJuLgABOAAsYMMOwBctAOZs5AVwBGAOhEhqIvqLkBDegGt21IsWqth29u2Bi7ANgDsAJgoRUXZ4EAtrJgAxYB4JZmIAZXo5JjC4AF44AB4EuCYADwlRYk5kfQAtEgAVAE8wJgBBURqAPgAKVTgEgEoMlsQKOCG4ZhhtKFEUfQsoKAsatrLtUWtRCAB3UTbu7v0YOeDMaBA2tot2XvT+hEHhu6CQkaZ2bViVLoBtAF0MuG-KO53I5QOBtB7wCxwCCYODnXo3QGIhieeBgGbsFi-dj6dgWAQABXRTDO3QBSLu-FBaKgGOIOO09HozwuA3J5OYL1i+jArjkbWptP0xAsMAspNubPQEsBaGld1G40mHNeMDJQzQ4rQlECKLgYGA9DGzCSKTSmTKEF0ACsmIa2gjhuMeO8yiEoMBROptgAaaXrYDEGByF36UTaEC6JhQH3SlLAdQKENhiNRmMaqjgvU4TAxJgm1KQ836S02u0OoaiCwgJght0er3dX13CzqGtTZOR6ON6X6w3jZ7vMIRaKxKMsfNhfkGo155IF7vpnXBeDEXS-C3W20we3S5iaMS1-b1mN3MDZ3OdIdRGJxcdzydn6G5idi32LmB1Jh66f99f6D0CFAWQfvU0Lfn2xr3hYLQUCBX6PjmPBfkWAFRsBn5gQhz5QTBcFwKuf6oUBcFgauMGwZ+IwWOs+I-swvzlnATrvHWnpNsM-qBsGcAdlG7FDHGCYwO8vFQG+VB4XMNHnkhDHSpW1YsUebHSi2baifx4Ezp0Um0RBTDfOJFH1FR6wACJrpkjF7h4ohKe6KmnjJA6mfizmGRQspLo8mD0L8bRSRZ7zLKsGyiN07yBZZ1y7kwRqTKu3JEgF1EWd0nkUEAA)

Here the same example using the version down below [link](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAbzgLzgXzgMyhEcDkyEAJvgNwBQFAxhAHYDO8IAhgNYCmAYsADYwcoHYgGVqACw6sA6sBjiACkLA5qHBgzgBeOAB4RcDgA8BdYpuQA6AFokAKgE8wHAIJ0HAPgAUDAFxwRAEptD0QKODghGABXKDoUSxUOFQg1DS8vFgZgrVCEcIi4WkZ4IQZo-n8RAG0AXW04OspCuGBMOC8AQhcoKBYHS2AGHr6HTOzg-JaIqNj4soqYZsK0AojMaA7ipjgWOAh2rMm1wu34MBYoBmEGhksGFkwOBUvrzMDllraOi6vhe+i1DS2TC00KC34iWiDHEXl+12IlmILBgLA+JwiqxaWPBHBicUi6kWyzQABoEpdRj5Aui0JQaPQdmBgNR8RwxJJWA0rBAAEYAKw4rK8U0KsV4-isTCgwDoAHMvIFSScAO7AYjySWWOjREC8wSK5UtSTAOXiGBanV6g1KihoQJUM5wFKYPjsiRSPY6HkCoUwEUnOgsEAcLXS2UK20tFhy0MJK36qCGk7M1mxdT+VicHj8QTCDme2TyJTJVTqBhwllsgusW32x2M+DEXncyx8wXC0UzDhy4D0MMwGXy5MtF1uvxwLPcPgCISiD0yOSKZRl9Jj3juzlo5X1igwJwcZ1V9Ot2VPKC6ffOA5HtNCGssDx7g-OnCujenujny8Hm-rzeek+V6Hs2n7fsBN7Nk+z7OJELAqgox5CA0XZwOK-jhvKRqFGqGriP4CaCNhEQmmaFpwIRUA7lQwFwQhb5uihgbBnGmFysRuyxgRuqJhxqZshOfQIUhHB1NRMGHkJAAiLY6KhQi9v2cBsXxDEboJ8FKAcbpiXaDYlFg1ANF40m8v40R0GwdAQCqdCBP4pkhKCuL4vEzaJK8HAmfBMkOmgQA)

```ts
const makeFilteredSchema = <S extends z.ZodTypeAny>(s: S) => {
  return z.preprocess((as) => {
    const result: S[] = [];
    if (!Array.isArray(as)) {
      return result;
    }
    for (const a of as) {
      const parsed = s.safeParse(a);
      if (parsed.success) {
        result.push(parsed.data);
      }
    }
    return result;
  }, z.array(s));
};
  ```
